streams/test-utils/src/main/java/org/apache/kafka/streams/TestOutputTopic.java

```diff
@@ -0,0 +1,200 @@
+ * <p>
+ * If you need to test key, value and headers, use {@link #readRecord()} methods.
+ * Using {@link #readKeyValue()} you get directly KeyValue, but have no access to headers any more
+ * Using {@link #readValue()} you get directly value, but have no access to key or  headers any more
+ *
+ * <h2>Processing messages</h2>
+ * <pre>{@code
+ *     private TestOutputTopic<String, Long> outputTopic;
+ *      ...
+ *     outputTopic = testDriver.createOutputTopic(OUTPUT_TOPIC, stringDeserializer, longDeserializer);
+ *     ...
+ *     assertThat(outputTopic.readValue()).isEqual(1);
+ * }</pre>
+ *
+ * @param <K> the type of the Kafka key
+ * @param <V> the type of the Kafka value
+ * @see TopologyTestDriver
+ */
+public class TestOutputTopic<K, V> {
+    private final TopologyTestDriver driver;
+    private final String topic;
+    private final Deserializer<K> keyDeserializer;
+    private final Deserializer<V> valueDeserializer;
+
+    /**
+     * Create a test output topic to read messages from
+     *
+     * @param driver            TopologyTestDriver to use
+     * @param topicName         the topic name used
+     * @param keyDeserializer   the key deserializer
+     * @param valueDeserializer the value deserializer
+     */
+    TestOutputTopic(final TopologyTestDriver driver,
+                    final String topicName,
+                    final Deserializer<K> keyDeserializer,
+                    final Deserializer<V> valueDeserializer) {
+        Objects.requireNonNull(driver, "TopologyTestDriver cannot be null");
+        Objects.requireNonNull(topicName, "topicName cannot be null");
+        Objects.requireNonNull(keyDeserializer, "keyDeserializer cannot be null");
+        Objects.requireNonNull(valueDeserializer, "valueDeserializer cannot be null");
+        this.driver = driver;
+        this.topic = topicName;
+        this.keyDeserializer = keyDeserializer;
+        this.valueDeserializer = valueDeserializer;
+    }
+
+
+    /**
+     * Read one Record from output topic and return value only.
+     * <p>
+     * Note. The key and header is not available
+     *
+     * @return Next value for output topic
+     */
+    public V readValue() {
+        final TestRecord<K, V> record = readRecord();
+        return record.value();
+    }
+
+    /**
+     * Read one Record KeyValue from output topic.
+     *
+     * @return Next output as KeyValue
+     */
+    public KeyValue<K, V> readKeyValue() {
+        final TestRecord<K, V> record = readRecord();
+        return new KeyValue<>(record.key(), record.value());
+    }
+
+    /**
+     * Read one Record from output topic.
+     *
+     * @return Next output as TestRecord
+     */
+    public TestRecord<K, V> readRecord() {
+        return driver.readRecord(topic, keyDeserializer, valueDeserializer);
+    }
+
+    /**
+     * Read output to List.
+     * If the existing key is modified, it can appear twice in output, but replaced in map
```

Comment:
I know the context of the KIP, but I think it's hard to understand for users what this means.

`This method can be used if the result is considered a stream. If the result is considered a table, the list will contain all updated, ie, a key might be contained multiple times. If you are only interested in the last table update (ie, the final table state), you can use {@link #readKeyValuesToMap()} instead.`

Reply:
changed